<h1 style="color:rgb(0,120,170)">Hands-on AI II</h1>
<h2 style="color:rgb(0,120,170)">Unit 9 (Assignment) -- Introduction to Reinforcement Learning -- Part II </h2>

<h2>Exercise 0</h2>

- Import the same modules as discussed in the lecture notebook.
- Check if your model versions are correct.

In [193]:
# your imports go here
import u9_utils as u9
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import numpy as np
import gym

from typing import Any, Dict, Tuple
from gym.envs.toy_text import TaxiEnv

# Set Seaborn plotting style.
sns.set()

import gym

from gym.envs.toy_text import FrozenLakeEnv 


In [194]:
# your modul version check goes here
u9.check_module_versions()


Installed Python version: 3.7 (✓)
Installed matplotlib version: 3.1.3 (✓)
Installed Pandas version: 1.0.3 (✓)
Installed Seaborn version: 0.10.1 (✓)
Installed OpenAI Gym version: 0.17.2 (✓)


All exercises in this assignment are referring to the <i>FrozenLake-v0</i> environment of <a href="https://gym.openai.com"><i>OpenAI Gym</i></a>. This environment is descibed according to its official <a href="https://gym.openai.com/envs/FrozenLake-v0/">OpenAI Gym website</a> as follows:<br>
<cite>Winter is here. You and your friends were tossing around a frisbee at the park when you made a wild throw that left the frisbee out in the middle of the lake. The water is mostly frozen, but there are a few holes where the ice has melted. If you step into one of those holes, you'll fall into the freezing water. At this time, there's an international frisbee shortage, so it's absolutely imperative that you navigate across the lake and retrieve the disc. However, the ice is slippery, so you won't always move in the direction you intend.</cite>


There are <i>four</i> types of surfaces described in this environment:
<ul>
    <li><code>S</code> $\rightarrow$ starting point (<span style="color:rgb(0,255,0)"><i>safe</i></span>)</li>
    <li><code>F</code> $\rightarrow$ frozen surface (<span style="color:rgb(0,255,0)"><i>safe</i></span>)</li>
    <li><code>H</code> $\rightarrow$ hole (<span style="color:rgb(255,0,0)"><i>fall to your doom</i></span>)</li>
    <li><code>G</code> $\rightarrow$ goal (<span style="color:rgb(255,0,255)"><i>frisbee location</i></span>)</li>
</ul>


If not already done, more information on how to <i>install</i> and <i>import</i> the <code>gym</code> module is available in the lecture's notebook.

<h3 style="color:rgb(0,120,170)">States and actions</h3>
Experiment with the <i>FrozenLake-v0</i> environment as discussed during the lecture and explained in the accompanying notebook.

In [195]:
lake_environment = FrozenLakeEnv()
u9.set_seed(environment=lake_environment, seed=42)


In [196]:
lake_environment.render(mode=r'human')
current_state_id = lake_environment.s
print(f'\nCurrent state ID: {current_state_id}')



SFFF
FHFH
FFFH
HFFG

Current state ID: 0


The current position of the <i>disc retrieving</i> entity is displayed as a filled <span style="color:rgb(255,0,0)"><i>red</i></span> rectangle.

As we want to tackle this problem using our renowned <i>random search</i> approach, we have to analyse its applicability beforehand. Hence, the number of possible <i>actions</i> and <i>states</i> is of utter importance, as we don't want to get lost in the depth of combinatorial explosion.
<ul>
    <li>Query the amount of <i>actions</i> using the appropriate peoperty of the lake environment.</li>
    <li>Query the amount of <i>states</i> using the appropriate property of the lake environment.</li>
</ul>

In [197]:
num_actions = lake_environment.action_space.n
num_states = lake_environment.observation_space.n
print(f'The FrozenLake-v0 environment comprises <{num_actions}> actions and <{num_states}> states.')


The FrozenLake-v0 environment comprises <4> actions and <16> states.


<h2>Exercise 1</h2>

- Create a q_table for the frozen lake environment.
- Apply $Q$-learning as it was done in the lecture to solve the environment.
- Test the learned policy and animate one (or more) exemplary episode.
- What do you observe? Does the agent learn anything useful? Discuss if something strange happens. Hint: print the q_table during training to better understand what is going on during learning.

In [198]:
# q_table
q_table = np.zeros([lake_environment.observation_space.n, lake_environment.action_space.n])
q_table
# lines: states
# columns: actions


array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [199]:
# shape of q_table
q_table.shape


(16, 4)

Problem: ordered action is not always done action. --> Next state can't be expected that easily. 


In [200]:
def apply_q_learning(environment: lake_environment, alpha: float = 0.1):
    """
    Solve lake_environment by applying Q learning.
    """
    for i in range(1, 10001):
        # your code goes here
        state = environment.reset()
        done = False
    
        while not done:
            action = np.argmax(q_table[state]) 
            next_state, reward, done, info = environment.step(action) 
            old_value = q_table[state, action] # old q value
            next_max = np.max(q_table[next_state])
            new_value = (1 - alpha) * old_value + alpha * (reward + next_max)
            q_table[state, action] = new_value # update Q table

            state = next_state
        
        # only for printing:
        if i % 100 == 0:
            clear_output(wait=True)
            print(f"Episode: {i}")
            print(q_table)


        # until here
    print("Training finished.\n")
    

In [201]:
%%time
from IPython.display import clear_output
# train your agent
alpha = 0.1
apply_q_learning(lake_environment, alpha)


Episode: 10000
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
Training finished.

CPU times: user 9.59 s, sys: 1.12 s, total: 10.7 s
Wall time: 8.88 s


In [202]:
total_epochs, total_dives = 0, 0
episodes = 100

captured_frames = [[] for _ in range(episodes)]

for episode in range(episodes):
    # test your method
    state = lake_environment.reset()
    epochs, penalties, reward = 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = lake_environment.step(action)

        if done and (reward == 0.0):
            penalties += 1 # when false drop off

        epochs += 1

        captured_frames[episode].append({
            r'frame': lake_environment.render(mode=r'ansi'),
            r'state': state,
            r'action': action,
            r'reward': reward
        })

    total_dives += penalties
    total_epochs += epochs

print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average dives per episode: {total_dives / episodes}")

Results after 100 episodes:
Average timesteps per episode: 17.19
Average dives per episode: 1.0


In [203]:
u9.animate_environment_search(frames=captured_frames[12], verbose=True, delay=0.4)


  (Left)
SFFF
FHFH
FFFH
HFFG

Step No.: 5
State ID: 12
Action ID: 0
Reward: 0.0


In [204]:
u9.animate_environment_search(frames=captured_frames[8], verbose=True, delay=0.4)


  (Left)
SFFF
FHFH
FFFH
HFFG

Step No.: 29
State ID: 12
Action ID: 0
Reward: 0.0


In [187]:
u9.animate_environment_search(frames=captured_frames[50], verbose=True, delay=0.4)


  (Left)
SFFF
FHFH
FFFH
HFFG

Step No.: 295
State ID: 5
Action ID: 0
Reward: 0.0


What do you observe? Does the agent learn anything useful? Discuss if something strange happens. Hint: print the q_table during training to better understand what is going on during learning. <br>
The q_table stays as it is only filled up with 0s. It follows that the model doesn't seem to learn anything. 
The problem is that the agent always acts with the same probabilities without changing them. It learns nothing.

<h2>Exercise 2</h2>
Very likely your training in Exercise 1 was not successful. Try to add exploration to your algorithm (you might have to write a new function):
<li><code>I</code> $\rightarrow$ Throw a random uniform number between 0 and 1. 
<li><code>II</code> $\rightarrow$ If the number is smaller than 0.1, sample a random action.
<li><code>III</code> $\rightarrow$ Choose your action as usual.   
    
- Apply the modified $Q$-learning again to solve the environment.
- Test the learned policy and animate one (or more) exemplary episode.
- What do you observe? Does the agent learn now?.

In [99]:
# q_table
del q_table
q_table = np.zeros([lake_environment.observation_space.n, lake_environment.action_space.n])
q_table


array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [178]:
import random
def apply_q_learning_exploration(environment: lake_environment, alpha: float = 0.1):
    """
    Solve lake_environment by applying Q learning and exploration.
    """
    for i in range(1, 10001):
        # your code goes here

        
        state = environment.reset()
        done = False
    
        while not done:
            
            # I: get a random number between 0 and 1
            rand = random.randint(0, 10000)
            number = rand/10000
            
            if number < 0.1:# II: make a random action when number smaller than 0.1
                action = environment.action_space.sample()
                
            else:# III: choose action as usual
                action = np.argmax(q_table[state]) 
                
            next_state, reward, done, info = environment.step(action) 
            old_value = q_table[state][action] # old q value
            next_max = np.max(q_table[next_state])
            new_value = (1 - alpha) * old_value + alpha * (reward + next_max)

            q_table[state, action] = new_value # update Q table

            state = next_state
            
        
        # only for printing:
        if i % 100 == 0:
            clear_output(wait=True)
            print(f"Episode: {i}")
            print(q_table)  
   
    print("Training finished.\n")
    

In [179]:
%%time
from IPython.display import clear_output
# train your agent
alpha = 0.1
apply_q_learning_exploration(lake_environment, alpha)


Episode: 10000
[[0.87874789 0.87759927 0.87607821 0.88044035]
 [0.50881911 0.58336085 0.62660669 0.88044035]
 [0.84435986 0.70403614 0.79603599 0.88044035]
 [0.54321723 0.69570303 0.60090239 0.88044035]
 [0.87196583 0.62799478 0.60722176 0.63168139]
 [0.         0.         0.         0.        ]
 [0.5922006  0.33148232 0.42363847 0.11253993]
 [0.         0.         0.         0.        ]
 [0.67027754 0.55595798 0.61912579 0.85930305]
 [0.56046815 0.84475158 0.6079083  0.52366454]
 [0.78855707 0.64004165 0.57295127 0.53995953]
 [0.         0.         0.         0.        ]
 [0.         0.         0.         0.        ]
 [0.56787153 0.80096881 0.89906889 0.43769242]
 [0.86196977 0.94129664 0.91389607 0.90193405]
 [0.         0.         0.         0.        ]]
Training finished.

CPU times: user 36.9 s, sys: 6.72 s, total: 43.7 s
Wall time: 32.8 s


In [164]:
total_epochs, total_dives = 0, 0
episodes = 100

captured_frames = [[] for _ in range(episodes)]

for episode in range(episodes):
    # test your method
    state = lake_environment.reset()
    epochs, penalties, reward = 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = lake_environment.step(action)

        if done and (reward == 0.0):
            penalties += 1 # when false drop off

        epochs += 1

        # Save rendering of current state.
        captured_frames[episode].append({
            r'frame': lake_environment.render(mode=r'ansi'),
            r'state': state,
            r'action': action,
            r'reward': reward
        })

    total_dives += penalties
    total_epochs += epochs
    
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average dives per episode: {total_dives / episodes}")


Results after 100 episodes:
Average timesteps per episode: 91.15
Average dives per episode: 0.13


In [180]:
u9.animate_environment_search(frames=captured_frames[4], verbose=True, delay=0.1)


  (Down)
SFFF
FHFH
FFFH
HFFG

Step No.: 32
State ID: 15
Action ID: 1
Reward: 1.0


In [182]:
u9.animate_environment_search(frames=captured_frames[40], verbose=True, delay=0.1)


  (Left)
SFFF
FHFH
FFFH
HFFG

Step No.: 20
State ID: 5
Action ID: 0
Reward: 0.0


In [184]:
u9.animate_environment_search(frames=captured_frames[8], verbose=True, delay=0.1)


  (Down)
SFFF
FHFH
FFFH
HFFG

Step No.: 126
State ID: 15
Action ID: 1
Reward: 1.0


What do you observe? Does the agent learn now?<br>
When you have a look at the q_table now you can discover an interesting fact. On all states which are holes or the goal the probability for all actions there is 0. That's because the game is over then. However, there is no probability of 0 for the remaining fields. <br>
The average of dives per episode is only 13 % now (not 100 % like before). Because of slippery ice it can happen that you fall in the water although you have acted in a right way. Therefore, the average dive probability is even smaller when the ice wouldn't be slippery.<br>
Furthermore, the agent learnt to connect special states with individual actions. 